include/deal.II/matrix_free/operators.h

```diff
@@ -834,6 +939,183 @@ namespace MatrixFreeOperators
+
+      return res;
+    }
+  }
+
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, int n_components, typename Number>
+  template <bool homogeneous>
+  void
+  LaplaceOperator<dim, fe_degree, n_q_points_1d, n_components, Number>::
+  local_apply_cell (const MatrixFree<dim,Number>                     &data,
+                    LinearAlgebra::distributed::Vector<Number>       &dst,
+                    const LinearAlgebra::distributed::Vector<Number> &src,
+                    const std::pair<unsigned int,unsigned int>  &cell_range) const
+  {
+    FEEvaluation<dim,fe_degree,n_q_points_1d,n_components,Number> phi (data);
+    for (unsigned int cell=cell_range.first; cell<cell_range.second; ++cell)
+      {
+        phi.reinit (cell);
+        phi.read_dof_values(src);
+        phi.evaluate (false,true,false);
+        for (unsigned int q=0; q<phi.n_q_points; ++q)
+          if (homogeneous)
+            {
+              phi.submit_gradient (phi.get_gradient(q), q);
+            }
+          else
+            {
+              Assert (non_negative((*scalar_coefficient)(cell,q)),
+                      ExcMessage("Coefficient must be non-negative"));
+              phi.submit_gradient ((*scalar_coefficient)(cell,q)*phi.get_gradient(q), q);
+            }
+        phi.integrate (false,true);
+        phi.distribute_local_to_global (dst);
+      }
+  }
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, int n_components, typename Number>
+  template <bool homogeneous>
+  void
+  LaplaceOperator<dim, fe_degree, n_q_points_1d, n_components, Number>::
+  local_diagonal_cell (const MatrixFree<dim,Number>                     &data,
+                       LinearAlgebra::distributed::Vector<Number>       &dst,
+                       const unsigned int &,
+                       const std::pair<unsigned int,unsigned int>       &cell_range) const
+  {
+    FEEvaluation<dim,fe_degree,n_q_points_1d,n_components,Number> phi (data);
+    for (unsigned int cell=cell_range.first; cell<cell_range.second; ++cell)
+      {
+        phi.reinit (cell);
+        VectorizedArray<Number> local_diagonal_vector[phi.tensor_dofs_per_cell];
+        for (unsigned int i=0; i<phi.dofs_per_cell; ++i)
+          {
+            for (unsigned int j=0; j<phi.dofs_per_cell; ++j)
+              phi.begin_dof_values()[j] = VectorizedArray<Number>();
+            phi.begin_dof_values()[i] = 1.;
+            phi.evaluate (false,true,false);
+            for (unsigned int q=0; q<phi.n_q_points; ++q)
+              if (homogeneous)
+                {
+                  phi.submit_gradient (phi.get_gradient(q), q);
+                }
+              else
+                {
+                  Assert (non_negative((*scalar_coefficient)(cell,q)),
+                          ExcMessage("Coefficient must be non-negative"));
+                  phi.submit_gradient ((*scalar_coefficient)(cell,q)*phi.get_gradient(q), q);
+                }
+            phi.integrate (false,true);
```

Comment:
One more remark: How about using a separate function
```
void do_operation_on_cell(FEEvaluation<dim,fe_degree,n_q_points_1d,n_components,Number> &phi) const
{
  phi.evaluate(false, true);
  ...
  phi.integrate(false, true);
}
```
This function would be the same for both the diagonal computation as well as the `local_apply`, keeping the interesting stuff in a single place.